snakefmt/formatter.py

```diff
@@ -296,9 +311,34 @@ def format_param(
         if param_list:
             val = f"f({val})"
             extra_spacing = 3
+
+        # get the index of the last character of the first docstring, if any
+        docstring_index = index_of_first_docstring(val)
+        docstring_line_index = None
+        if docstring_index is not None:
+            docstring_line_index = val[:docstring_index].count("\n")
+        lines = val.splitlines()
+        if docstring_line_index is not None and docstring_line_index + 1 < len(lines):
+            docstring_has_extra_newline_after = (
+                lines[docstring_line_index + 1].strip() == ""
+            )
+        else:
+            docstring_has_extra_newline_after = False
+
         val = self.run_black_format_str(
             val, target_indent, extra_spacing, no_nesting=True
         )
+
+        # remove newline added after first docstring (black>=24.1)
+        if docstring_line_index is not None and not docstring_has_extra_newline_after:
+            lines = val.splitlines()
+            if docstring_line_index + 1 < len(lines):
+                line_after_docstring = lines[docstring_line_index + 1]
+                if line_after_docstring.strip() == "":
+                    # delete the newline
+                    lines.pop(docstring_line_index + 1)
+                    val = "\n".join(lines)
```

Comment:
Let me try and recap, here, all of this code include function `index_of_first_docstring` is intended to remove black's addition of a newline before a docstring, right?

So, could we tell black somehow that this is a 'class/function-like' docstring (no need for newline addition), and not a 'module docstring'?

And if not, maybe a shorter (code-wise) option is to pass to black the docstring as if it were inside a class? (E.g. `class x:` before the docstring, passing to black, and removing `class x:`)
(I can give this a go if you think it's reasonable.)

Reply:
Yeah, it adds a space after the first module string. I did think about adding it inside a class or function as you say, but that would likely create *more* code and thought this was probably more succinct. Happy for you to refactor though if you want.

Reply:
Ok, I will try and have a look at this, but I've got holidays coming up - can have a look at the end of February. In the meantime this is released as is and that's good :+1: